### ISSUE TITLE
Extend nn.Module() with Support for Federated Learning of non-parameter information

### ISSUE THREAD

**Issue by iamtrask** [OpenMined/PySyft#3236]
**Is your feature request related to a problem? Please describe.**

https://github.com/OpenMined/PySyft/issues/2498

**Describe the solution you'd like**

Solution: extend nn.Module() with a .parameters_and_statistics() iterator which will look for this kind of information so that we can make sure that federated learning properly moves everything from machien to machine.

**Describe alternatives you've considered**
We could also expose .statistics() different from .parameters() so that federaetd learning optimizers could treat them differnetly

**Additional context**
It's important to consider that FL averaging might not work for all statistics like this.

**Comment by SagarMoghe** [comment]
Can I take up this problem?

**Comment by AshwinRam2000** [comment]
I can take this up.

**Comment by karlhigley** [comment]
Assigned you both. Hope that whoever starts working on it first can post a WIP PR with their early commits so you can collaborate on it.

**Comment by SagarMoghe** [comment]
Thank you

**Comment by ratmcu** [comment]
@karlhigley Can I work on this issue as well. I have already worked on a use case similar. So we can collaboratively solve this.

**Comment by SagarMoghe** [comment]
@iamtrask @karlhigley , I am new to Syft so can you please guide me where in the code base should I start to look for relevant code to the related issue. Thank you

**Comment by sachin-101** [comment]
@iamtrask From the comments in #2498 it seems the BatchNorm bug was due to different versions. So you can close this issue if it is solved.

**Comment by karlhigley** [comment]
I think the proposed feature is still useful.

**Comment by SagarMoghe** [comment]
@iamtrask @karlhigley  I did some research and I went around and snooped into nn.module code and noticed that line 87 of [nn.module](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py) has the variable that contains information I need. I am not sure where exactly in the pysyft codebase should I implement the function to get the value of this variable. I am new to pysyft and dont know much about the code structure. Please help me.
PS: does making a function .parameters_and_statistics() which returns the self._buffers containing all the non parameterised attributes of the neuralNet and returning them to the user solves the problem of the issue?
Thank you

**Comment by ratmcu** [comment]
@SagarMoghe I already added a function and working on testing it,
I am also adding features to get all the non parameter elements of iterated by an arbitrary function that could be added to the hooked pytorch nn module to be considered at the send().
I would like if we can work on testing the new feature, rather than repeating the same.

**Comment by sachin-101** [comment]
@ratmcu You can start working on a new branch and push the WIP code here. That might be a good way to collaborate.

**Comment by johnnylin110** [comment]
Hello , I am the one who comment in the issue #2498 
And I found it might have still a issue their . So I comment below, and the #2498 is closed so I mention in here , maybe need to take a look?

**Comment by SagarMoghe** [comment]
They were using an old pytorch version and had the issue but the functionality suggested is still useful.

**Comment by ratmcu** [comment]
added a PR https://github.com/OpenMined/PySyft/pull/3273
